File: cpp/src/Thermostat.cpp
#include "Thermostat.hpp"

Thermostat::Thermostat(int id, const std::string &name, const std::string &location)
    : Device(id, name, location), targetTemp(20.0), mode(0) {}

Thermostat::Thermostat(const Thermostat& other)
    : Device(other), targetTemp(other.targetTemp), mode(other.mode) {}

Thermostat::~Thermostat() {}

void Thermostat::setTargetTemp(double t) {
    targetTemp = t;
    std::cout << "Термостат '" << getName() << "' установлена температура: " << targetTemp << std::endl;
}

void Thermostat::setMode(int m) {
    mode = m;
    std::cout << "Термостат '" << getName() << "' установлен режим: " << mode << std::endl;
}

void Thermostat::autoAdjust() {
    std::cout << "Термостат '" << getName() << "' авторегулировка (демо)" << std::endl;
}

void Thermostat::showStatus() {
    std::cout << "Термостат '" << getName() << "' целевая температура=" << targetTemp << " режим=" << mode << std::endl;
}

File: java/Thermostat.java
package smarthome;

public class Thermostat extends AbstractDevice {

    private double targetTemp;
    private String mode;

    public Thermostat(String id, String name, String location, int powerUsage, double targetTemp) {
        super(id, name, location, powerUsage);
        this.targetTemp = targetTemp;
        this.mode = "auto";
    }

    public void setTargetTemp(double t) {
        targetTemp = t;
        System.out.println(name + ": цель " + targetTemp + "°C");
    }

    public void setMode(String m) {
        mode = m;
        System.out.println(name + ": режим " + mode);
    }

    public void perform(String command) {
        try {
            String c = command.toLowerCase();
            if (c.startsWith("вкл") || c.startsWith("turnon")) {
                turnOn();
            } else if (c.startsWith("выкл") || c.startsWith("turnoff")) {
                turnOff();
            } else if (c.startsWith("темп") || c.startsWith("temp")) {
                try {
                    String num = command.replaceAll("[^0-9.,-]", "").trim();
                    if (num.isEmpty()) throw new NumberFormatException("no number");
                    setTargetTemp(Double.parseDouble(num));
                } catch (NumberFormatException e) {
                    System.out.println(name + ": неверная температура");
                }
            } else {
                System.out.println(name + ": неизвестная команда для термостата");
            }
        } catch (Exception e) {
            System.out.println(name + ": ошибка выполнения команды термостата: " + e.getMessage());
        }
    }

    @Override
    public String toString() {
        return super.toString() + " цель=" + targetTemp + "C";
    }
}
